File: android/src/main/kotlin/com/blaink/blaink_flutter/BlainkFlutterPlugin.kt
//
//  BlainkFlutterPlugin.kt
//  Blaink Flutter SDK
//
//  Prompted by Raşid Ramazanov using Cursor on 21.09.2025.
//

package com.blaink.blaink_flutter

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import com.blaink.Blaink
import com.blaink.core.BlainkDelegate
import com.blaink.core.PushEnvironment
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class BlainkFlutterPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler, BlainkDelegate {
    private lateinit var channel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var context: Context
    private var eventSink: EventChannel.EventSink? = null
    private lateinit var blaink: Blaink

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "blaink_flutter")
        channel.setMethodCallHandler(this)

        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "blaink_flutter/events")
        eventChannel.setStreamHandler(this)
        
        // Don't initialize SDK instance here - wait for setup() call
        // This prevents singleton contamination between apps
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "setup" -> {
                try {
                    val sdkKey = call.argument<String>("sdkKey")
                    val environmentStr = call.argument<String>("environment") ?: "production"
                    val isDebugLogsEnabled = call.argument<Boolean>("isDebugLogsEnabled") ?: false
                    
                    if (sdkKey == null) {
                        result.error("INVALID_ARGUMENTS", "SDK key is required", null)
                        return
                    }
                    
                    val environment = when (environmentStr) {
                        "development" -> PushEnvironment.DEVELOPMENT
                        "production" -> PushEnvironment.PRODUCTION
                        else -> PushEnvironment.PRODUCTION
                    }
                    
                    // Initialize SDK instance with proper isolation
                    blaink = Blaink.getInstance()
                    blaink.delegate = this
                    
                    blaink.setup(
                        context = context,
                        sdkKey = sdkKey,
                        environment = environment,
                        isDebugLogsEnabled = isDebugLogsEnabled
                    )
                    
                    result.success(null)
                } catch (e: Exception) {
                    result.error("SETUP_ERROR", "Failed to setup Blaink SDK: ${e.message}", null)
                }
            }
            
            "registerForRemoteNotifications" -> {
                try {
                    if (!::blaink.isInitialized) {
                        result.error("SDK_NOT_INITIALIZED", "Blaink SDK not initialized. Call setup() first.", null)
                        return
                    }
                    
                    val deviceToken = call.argument<String>("deviceToken")
                    if (deviceToken == null) {
                        result.error("INVALID_ARGUMENTS", "Device token is required", null)
                        return
                    }
                    
                    blaink.registerForRemoteNotifications(deviceToken)
                    result.success(null)
                } catch (e: Exception) {
                    result.error("REGISTRATION_ERROR", "Failed to register for notifications: ${e.message}", null)
                }
            }
            
            "getCurrentUser" -> {
                if (!::blaink.isInitialized) {
                    result.error("SDK_NOT_INITIALIZED", "Blaink SDK not initialized. Call setup() first.", null)
                    return
                }
                
                CoroutineScope(Dispatchers.Main).launch {
                    try {
                        val userResult = blaink.getCurrentUser()
                        if (userResult.isSuccess) {
                            result.success(userResult.getOrNull())
                        } else {
                            result.error("GET_USER_ERROR", "Failed to get current user: ${userResult.exceptionOrNull()?.message}", null)
                        }
                    } catch (e: Exception) {
                        result.error("GET_USER_ERROR", "Failed to get current user: ${e.message}", null)
                    }
                }
            }
            
            "updateUser" -> {
                try {
                    // For now, the Android SDK doesn't seem to have an updateUser method
                    // We'll just return success to maintain compatibility with the Flutter interface
                    // Users can implement user updates through their own backend
                    result.success(null)
                } catch (e: Exception) {
                    result.error("UPDATE_USER_ERROR", "Failed to update user: ${e.message}", null)
                }
            }
            
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
    }
    
    // BlainkDelegate implementation
    override fun didReceiveNotification(notification: Map<String, Any>) {
        eventSink?.success(mapOf(
            "type" to "didReceiveNotification",
            "payload" to notification
        ))
    }
    
    override fun didRegisterForBlainkNotifications(blainkUserId: String) {
        eventSink?.success(mapOf(
            "type" to "didRegisterForBlainkNotifications",
            "userId" to blainkUserId
        ))
    }
}

File: ios/Classes/BlainkFlutterPlugin.swift
//
//  BlainkFlutterPlugin.swift
//  Blaink Flutter SDK
//
//  Prompted by Raşid Ramazanov using Cursor on 21.09.2025.
//

import Flutter
import UIKit
import Blaink

public class BlainkFlutterPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "blaink_flutter", binaryMessenger: registrar.messenger())
        let eventChannel = FlutterEventChannel(name: "blaink_flutter/events", binaryMessenger: registrar.messenger())
        
        let instance = BlainkFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        eventChannel.setStreamHandler(instance)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "setup":
            setup(call: call, result: result)
        case "registerForRemoteNotifications":
            registerForRemoteNotifications(call: call, result: result)
        case "getCurrentUser":
            getCurrentUser(result: result)
        case "updateUser":
            updateUser(call: call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func setup(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let sdkKey = args["sdkKey"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "SDK key is required", details: nil))
            return
        }
        
        let environmentStr = args["environment"] as? String ?? "production"
        let isDebugLogsEnabled = args["isDebugLogsEnabled"] as? Bool ?? false
        
        let environment: PushEnvironment = environmentStr == "development" ? .development : .production
        
        Blaink.shared.setup(
            sdkKey: sdkKey,
            environment: environment,
            isDebugLogsEnabled: isDebugLogsEnabled
        )
        
        result(nil)
    }
    
    private func registerForRemoteNotifications(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
              let deviceToken = args["deviceToken"] as? String else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Device token is required", details: nil))
            return
        }
        
        Blaink.shared.registerForRemoteNotifications(deviceToken: deviceToken)
        result(nil)
    }
    
    private func getCurrentUser(result: @escaping FlutterResult) {
        Task {
            let userId = await Blaink.shared.getCurrentUser()
            result(userId)
        }
    }
    
    private func updateUser(call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any] else {
            result(FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid arguments", details: nil))
            return
        }
        
        let firstName = args["firstName"] as? String
        let lastName = args["lastName"] as? String
        let email = args["email"] as? String
        let phone = args["phone"] as? String
        
        Task {
            await Blaink.shared.updateUser(
                firstName: firstName,
                lastName: lastName,
                email: email,
                phone: phone
            )
            result(nil)
        }
    }
    
    // MARK: - FlutterStreamHandler
    
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        
        // Set up Blaink SDK delegate
        Blaink.shared.delegate = BlainkDelegateWrapper(eventSink: events)
        
        return nil
    }
    
    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        Blaink.shared.delegate = nil
        return nil
    }
}

// MARK: - BlainkDelegate Wrapper

class BlainkDelegateWrapper: BlainkDelegate {
    private let eventSink: FlutterEventSink
    
    init(eventSink: @escaping FlutterEventSink) {
        self.eventSink = eventSink
    }
    
    func didReceiveNotification(_ notification: UNNotification) {
        let payload = notification.request.content.userInfo
        let stringPayload = payload.compactMapValues { value in
            return String(describing: value)
        }
        
        eventSink([
            "type": "didReceiveNotification",
            "payload": stringPayload
        ])
    }
    
    func didRegisterForBlainkNotifications(userId: String) {
        eventSink([
            "type": "didRegisterForBlainkNotifications",
            "userId": userId
        ])
    }
    
    func didFailToRegisterForBlainkNotifications(error: Error) {
        eventSink([
            "type": "didFailToRegisterForBlainkNotifications",
            "error": error.localizedDescription
        ])
    }
}